File: app/src/main/java/com/example/anmol/beacons/MainActivity.java
package com.example.anmol.beacons;

import android.Manifest;
import android.annotation.TargetApi;
import android.content.DialogInterface;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
//import android.support.v4.app.Fragment;
//import android.support.v4.app.FragmentManager;
//import android.support.v4.app.FragmentPagerAdapter;
//import android.support.v7.app.AlertDialog;
//import android.support.v7.app.AppCompatActivity;
//import android.support.v7.widget.Toolbar;


import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.example.anmol.beacons.BeaconSearch.BeaconSearc;
import com.example.anmol.beacons.BeaconSimulator.BeaconSimu;
import com.google.android.material.tabs.TabLayout;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    // ToolBar
    private Toolbar toolbar;

    // TabLayout
    private TabLayout tabLayout;

    // ViewPager
    private ViewPager viewPager;


    private static final int PERMISSION_REQUEST_COARSE_LOCATION = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Checking Permission whether ACCESS_COARSE_LOCATION permssion is granted or not
        checkPermission();

        // Setting up of view
        setContentView(R.layout.activity_main);

        // Intializing of Layout Views
        initializedLayout();
    }


    //Intializing of Layout Views
    public void initializedLayout(){

        // Setting up of customized toolbar
        toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        // Setting up of View Pager
        viewPager = (ViewPager) findViewById(R.id.viewpager);
        setupViewPager(viewPager);

        //Setting up of TabLayout
        tabLayout = (TabLayout) findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(viewPager);
    }



    // Checking Permission whether ACCESS_COARSE_LOCATION permssion is granted or not
    public void checkPermission(){

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M){
            if (this.checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION)!=PackageManager.PERMISSION_GRANTED){
                AlertDialog.Builder builder=new AlertDialog.Builder(this);
                builder.setTitle("Location Permission");
                builder.setPositiveButton("OK",null);
                builder.setOnDismissListener(new DialogInterface.OnDismissListener() {
                    @TargetApi(Build.VERSION_CODES.M)
                    @Override
                    public void onDismiss(DialogInterface dialogInterface) {
                        requestPermissions(new String[]{Manifest.permission.ACCESS_COARSE_LOCATION},PERMISSION_REQUEST_COARSE_LOCATION);
                    }
                });
                builder.show();
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           String permissions[], int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        switch (requestCode) {
            case PERMISSION_REQUEST_COARSE_LOCATION: {

                // If Permission is Granted than its ok
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {

                }

                // If not Granted then alert the user by the message
                else {
                    final AlertDialog.Builder builder = new AlertDialog.Builder(this);
                    builder.setTitle("Functionality limited");
                    builder.setMessage("Since location access has not been granted, this app will not be able to discover beacons when in the background.");
                    builder.setPositiveButton(android.R.string.ok, null);
                    builder.setOnDismissListener(new DialogInterface.OnDismissListener() {
                        @Override
                        public void onDismiss(DialogInterface dialog) {
                            checkPermission();
                        }
                    });
                    builder.show();
                }
                return;
            }
        }
    }

    /*
       Setting up of View Pager
       2 Fragments :
       -> BeaconSearc() == Search
       -> BeaconSimu() == Simulator
     */
    private void setupViewPager(ViewPager viewPager) {
        ViewPagerAdapter adapter = new ViewPagerAdapter(getSupportFragmentManager());
        adapter.addFragment(new BeaconSearc() , "Search");
        adapter.addFragment(new BeaconSimu() , "Simulator");
        viewPager.setAdapter(adapter);
    }


    // Setting up the ViewPagerAdapter
    class ViewPagerAdapter extends FragmentPagerAdapter {
        private final List<Fragment> mFragmentList = new ArrayList<>();
        private final List<String> mFragmentTitleList = new ArrayList<>();

        public ViewPagerAdapter(FragmentManager manager) {
            super(manager);
        }

        @Override
        public Fragment getItem(int position) {
            return mFragmentList.get(position);
        }

        @Override
        public int getCount() {
            return mFragmentList.size();
        }

        public void addFragment(Fragment fragment, String title) {
            mFragmentList.add(fragment);
            mFragmentTitleList.add(title);
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return mFragmentTitleList.get(position);
        }
    }

}

File: app/src/main/java/com/example/anmol/beacons/MainActivity.kt
package com.example.anmol.beacons

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.view.View
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import com.example.anmol.beacons.BeaconSearch.BeaconSearch
import com.example.anmol.beacons.beaconSimulator.BeaconListFragment
import com.example.anmol.beacons.mqtt.PublishFragment
import com.google.android.material.tabs.TabLayout

class MainActivity : AppCompatActivity() {

    // ToolBar
    private var toolbar: Toolbar? = null

    // TabLayout
    private var tabLayout: TabLayout? = null

    // ViewPager
    private var viewPager: ViewPager? = null
    @RequiresApi(Build.VERSION_CODES.Q)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Setting up of view
        setContentView(R.layout.activity_main)

        // Intializing of Layout Views
        initializedLayout()

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            askPermissions(true)
        }
    }

    //Intializing of Layout Views
    fun initializedLayout() {

        // Setting up of customized toolbar
        toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        // Setting up of View Pager
        viewPager = findViewById<View>(R.id.viewpager) as ViewPager
        setupViewPager(viewPager)

        //Setting up of TabLayout
        tabLayout = findViewById<View>(R.id.tabs) as TabLayout
        tabLayout!!.setupWithViewPager(viewPager)
    }


private fun askPermissions(isForOpen: Boolean) {
    isRationale = false
    val permissionsRequired: MutableList<String> = ArrayList()
    val permissionsList: MutableList<String> = ArrayList()
    if (!checkPermission(
            permissionsList,
            Manifest.permission.ACCESS_COARSE_LOCATION
        )
    ) permissionsRequired.add("Write External Storage")
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {

        if (!checkPermission(
                permissionsList,
                Manifest.permission.ACCESS_BACKGROUND_LOCATION
            )
        ) permissionsRequired.add("Write Contact")
    }
    if (permissionsList.size > 0 && !isRationale) {
//            if (permissionsRequired.size > 0) {
//            }
        if (isForOpen) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                ActivityCompat.requestPermissions(
                    this, permissionsList.toTypedArray(),
                    permissionRequest
                )
            }
        }
    } else if (isRationale) {
        if (isForOpen) {
            androidx.appcompat.app.AlertDialog.Builder(this)
                .setTitle("Permission Alert")
                .setMessage("You need to grant permissions manually. Go to permission and grant all permissions.")
                .setPositiveButton("OK") { dialog, _ ->
                    dialog.dismiss()
                    val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
                    val uri = Uri.fromParts(
                        "package",
                        packageName, null
                    )
                    intent.data = uri
                    startActivityForResult(intent, permissionRequestDialog)
                }
                .show()
        }
    } else {
//            startActivity(Intent(this@PermissionsActivity, SplashActivity::class.java))
//            finish()
//            removeContact()
    }
}

private fun checkPermission(permissionsList: MutableList<String>, permission: String): Boolean {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        if (ContextCompat.checkSelfPermission(
                this,
                permission
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            permissionsList.add(permission)
            // Check for Rationale Option
            if (!isFirst) {
                if (!ActivityCompat.shouldShowRequestPermissionRationale(this, permission)) {
                    isRationale = true
                    return false
                }
            }
        }
    }
    return true
}


override fun onRequestPermissionsResult(
    requestCode: Int,
    permissions: Array<String>,
    grantResults: IntArray
) {
    when (requestCode) {
        permissionRequest -> {
            val perms: MutableMap<String, Int> = HashMap()
            // Initial
            perms[Manifest.permission.ACCESS_COARSE_LOCATION] = PackageManager.PERMISSION_GRANTED
            perms[Manifest.permission.ACCESS_BACKGROUND_LOCATION] = PackageManager.PERMISSION_GRANTED
            // Fill with results
            var i = 0
            while (i < permissions.size) {
                perms[permissions[i]] = grantResults[i]
                i++
            }
            // Check for ACCESS_FINE_LOCATION
            if (perms[Manifest.permission.ACCESS_COARSE_LOCATION] == PackageManager.PERMISSION_GRANTED &&
                perms[Manifest.permission.ACCESS_BACKGROUND_LOCATION] == PackageManager.PERMISSION_GRANTED
            ) {
                // All Permissions Granted
//                    startActivity(Intent(this@PermissionsActivity, SplashActivity::class.java))
//                    finish()
//                    removeContact()
            } else {
                // Permission Denied
                Toast.makeText(this, "Some Permission is Denied.", Toast.LENGTH_SHORT)
                    .show()
                isFirst = false
                askPermissions(true)
            }
        }
        else -> super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }
}


    private fun setupViewPager(viewPager: ViewPager?) {
        val adapter = ViewPagerAdapter(supportFragmentManager)
        adapter.addFragment(BeaconSearch(), "Search")
        adapter.addFragment(BeaconListFragment(), "Simulator")
        adapter.addFragment(PublishFragment(), "Publish")
        viewPager!!.adapter = adapter
    }

    // Setting up the ViewPagerAdapter
    internal inner class ViewPagerAdapter(manager: FragmentManager?) : FragmentPagerAdapter(
        manager!!
    ) {
        private val mFragmentList: MutableList<Fragment> = ArrayList()
        private val mFragmentTitleList: MutableList<String> = ArrayList()
        override fun getItem(position: Int): Fragment {
            return mFragmentList[position]
        }

        override fun getCount(): Int {
            return mFragmentList.size
        }

        fun addFragment(fragment: Fragment, title: String) {
            mFragmentList.add(fragment)
            mFragmentTitleList.add(title)
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return mFragmentTitleList[position]
        }
    }

    companion object {
        private const val PERMISSION_REQUEST_COARSE_LOCATION = 1
        private const val PERMISSION_REQUEST_COARSE_LOCATION_BACK = 2

        private val permissionRequest=11
        private val permissionRequestDialog=123

        private var isRationale:Boolean=false
        private var isFirst:Boolean=true

    }
}